File: candy_distribution.py
# Approach - In line
# Time - O(N)
# Space - O(N)

class Solution:

    def candy(self, ratings: List[int]) -> int:
        
        if not ratings or len(ratings) == 0:
            return 0
        
        result = [1] * len(ratings) ##c initially assign one candy for all children

        # left pass, start from 1 inorder to compare with left neighbor
        for i in range(1, len(ratings)):
            if ratings[i] > ratings[i-1]:
                result[i] = result[i-1] + 1 # give one more than the neighbor not add one to itself
        
        # right pass
        for i in range(len(ratings)-2, -1, -1):
            if ratings[i] > ratings[i+1]:
                result[i] = max(result[i], result[i+1]+1)
        
        return sum(result)

File: Candy_distribution.java
class Solution {
    public int candy(int[] ratings) {
        
        if(ratings == null || ratings.length == 0) return 0;
        int [] candies = new int[ratings.length];
        Arrays.fill(candies, 1);
        
        //left pass
        for(int i = 1; i < ratings.length; i++){
            if(ratings[i] > ratings[i-1]) {
                candies[i] = candies[i-1] + 1;
            }
        }
        // int sum = candies[candies.length-1];
        //right pass
        for(int i =ratings.length-2; i >= 0; i--){
            if(ratings[i] > ratings[i+1]) {
                candies[i] = Math.max(candies[i], candies[i+1] + 1);
            }
        }
        int sum = 0;
        for(int ele: candies){
            sum += ele;
        }
        return sum;

    }
}

//TC: o(n)
//SC: o(n)
